core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -18,153 +18,173 @@ package zio
+        case Right(_)    => 0L
+      })
+
+    def withPermit[R, E, A](zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, A] =
+      withPermits(1L)(zio)
+
+    def withPermitScoped(implicit trace: Trace): ZIO[Scope, Nothing, Unit] =
+      withPermitsScoped(1L)
+
+    def withPermits[R, E, A](n: Long)(zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, A] =
+      ZIO.acquireReleaseWith(reserve(n))(_.release)(_.acquire *> zio)
+
+    def withPermitsScoped(n: Long)(implicit trace: Trace): ZIO[Scope, Nothing, Unit] =
+      ZIO.acquireRelease(reserve(n))(_.release).flatMap(_.acquire)
+
+    def releaseN(n: Long)(implicit trace: Trace): UIO[Unit] =
+      ZIO.succeed(releaseInternal(n))
+
+    private def reserve(n: Long)(implicit trace: Trace): UIO[Reservation] = {
+      if (n <= 0) ZIO.succeedNow(Reservation(ZIO.unit, ZIO.unit))
+      else Promise.make[Nothing, Unit].flatMap { promise =>
+        ZIO.succeed {
+          var loop = true
+          var result: Reservation = null
+          while (loop) {
+            val current = state.get()
+            current match {
+              case Right(permits) if permits >= n =>
+                if (state.compareAndSet(current, Right(permits - n))) {
+                  result = Reservation(ZIO.unit, ZIO.succeed(releaseInternal(n)))
+                  loop = false
+                }
+              case Right(permits) =>
+                val newQueue = ScalaQueue(promise -> (n - permits))
+                if (state.compareAndSet(current, Left(newQueue))) {
+                  result = Reservation(promise.await, ZIO.succeed(releaseInternal(n)))
+                  loop = false
                 }
-            case Right(permits) => ZIO.unit -> Right(permits + n)
-          }.flatten
-
-        def releaseN(n: Long)(implicit trace: Trace): UIO[Any] = {
-
-          @tailrec
-          def loop(
-            n: Long,
-            state: Either[ScalaQueue[(Promise[Nothing, Unit], Long)], Long],
-            acc: UIO[Any]
-          ): (UIO[Any], Either[ScalaQueue[(Promise[Nothing, Unit], Long)], Long]) =
-            state match {
-              case Right(permits) => acc -> Right(permits + n)
               case Left(queue) =>
-                queue.dequeueOption match {
-                  case None => acc -> Right(n)
-                  case Some(((promise, permits), queue)) =>
-                    if (n > permits)
-                      loop(n - permits, Left(queue), acc *> promise.succeedUnit)
-                    else if (n == permits)
-                      (acc *> promise.succeedUnit) -> Left(queue)
-                    else
-                      acc -> Left((promise -> (permits - n)) +: queue)
+                val newQueue = queue.enqueue(promise -> n)
+                if (state.compareAndSet(current, Left(newQueue))) {
+                  result = Reservation(promise.await, ZIO.succeed(releaseInternal(n)))
+                  loop = false
                 }
             }
```

Comment:
It would probably be good to call `Thread.onSpinWait()` to tell the compiler/JVM that we are doing a spin loop here.